File: 101-symmetric-tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public static void main(String[] args) {
        
        TreeNode root = new TreeNode(1, new TreeNode(2), new TreeNode(3));


        System.out.println("------------------");
        System.out.println(isSymmetric(root));
    }


    public static boolean isSymmetric(TreeNode root) {
        return isMirror(root, root);
    }
    
    
    public static boolean isMirror(TreeNode first, TreeNode second) {
        if (first == null && second == null) {
            return true;
        }

        if (first != null && second != null &&
                first.val == second.val) {
            return  (isMirror(first.left, second.right)
                && isMirror(first.right, second.left));
        }

        return false;
    }
}

File: 101-symmetric-tree/Solution.py


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.isMirror(root, root)
    
    def isMirror(self, left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:

        if not left or not right:
            return left == right
        
        if left.val == right.val:
            return self.isMirror(left.left, right.right) and self.isMirror(left.right, right.left)
        
        return False
    
root = TreeNode(
    1,
    TreeNode(2),
    TreeNode(2)
)

obj = Solution()

print(obj.isSymmetric(root))